plugins/net_plugin/net_plugin.cpp

```diff
@@ -1238,67 +1301,128 @@ namespace eosio {
-      constexpr size_t header_size = sizeof( payload_size );
-      static_assert( header_size == message_header_size, "invalid message_header_size" );
-      const size_t buffer_size = header_size + payload_size;
+      template< typename T>
+      static send_buffer_type create_send_buffer( uint32_t which, const T& v ) {
+         // match net_message static_variant pack
+         const uint32_t which_size = fc::raw::pack_size( unsigned_int( which ) );
+         const uint32_t payload_size = which_size + fc::raw::pack_size( v );
 
-      auto send_buffer = std::make_shared<vector<char>>( buffer_size );
-      fc::datastream<char*> ds( send_buffer->data(), buffer_size );
-      ds.write( header, header_size );
-      fc::raw::pack( ds, unsigned_int( which ) );
-      fc::raw::pack( ds, v );
+         const char* const header = reinterpret_cast<const char* const>(&payload_size); // avoid variable size encoding of uint32_t
+         constexpr size_t header_size = sizeof( payload_size );
+         static_assert( header_size == message_header_size, "invalid message_header_size" );
+         const size_t buffer_size = header_size + payload_size;
 
-      return send_buffer;
-   }
+         auto send_buffer = std::make_shared<vector<char>>( buffer_size );
+         fc::datastream<char*> ds( send_buffer->data(), buffer_size );
+         ds.write( header, header_size );
+         fc::raw::pack( ds, unsigned_int( which ) );
+         fc::raw::pack( ds, v );
 
-   static std::shared_ptr<std::vector<char>> create_send_buffer( const signed_block_ptr& sb ) {
-      // this implementation is to avoid copy of signed_block to net_message
-      // matches which of net_message for signed_block
-      fc_dlog( logger, "sending block ${bn}", ("bn", sb->block_num()) );
-      return create_send_buffer( signed_block_which, *sb );
-   }
+         return send_buffer;
+      }
+
+   };
 
-   static std::shared_ptr<std::vector<char>> create_send_buffer( const packed_transaction& trx ) {
-      // this implementation is to avoid copy of packed_transaction to net_message
-      // matches which of net_message for packed_transaction
-      return create_send_buffer( packed_transaction_which, trx );
+   struct block_buffer_factory : public buffer_factory {
+
+      /// caches result for subsequent calls, only provide same signed_block_ptr instance for each invocation.
+      const send_buffer_type& get_send_buffer( const signed_block_ptr& sb ) {
+         if( !send_buffer ) {
+            send_buffer = create_send_buffer( sb );
+         }
+         return send_buffer;
+      }
+
+   private:
+
+      static std::shared_ptr<std::vector<char>> create_send_buffer( const signed_block_ptr& sb ) {
+         static_assert( signed_block_which == fc::get_index<net_message, signed_block>() );
+         // this implementation is to avoid copy of signed_block to net_message
```

Comment:
How does this avoid a copy?  `create_send_buffer()` still calls `fc::raw::pack()` with the datastream and the message, which eventually calls datastream's `write()` method which calls `memcpy()`.

Reply:
`net_message` is a `variant< .., signed_block, ..>`. To call `create_send_buffer( const net_message& m )` you would need to construct a `net_message` which would require copying the `signed_block` into the `variant`. 
Here we call `create_send_buffer( uint32_t which, const T& v )` which creates the `vector<char>` from the `which` and the `signed_block` without creating a `net_message`.